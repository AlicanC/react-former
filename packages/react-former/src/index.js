// @flow

export { default as FormHelper } from './FormHelper';
export type { Props as FormHelperProps } from './FormHelper';
export { default as createForm } from './createForm';
export type { Form } from './createForm';
export type { FormState } from './createFormStore';
