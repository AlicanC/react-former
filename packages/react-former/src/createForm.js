// @flow

import createFormStore, {
  type FormStore,
  // type FormState,
  type FieldState,
  type FieldValidationResult,
} from './createFormStore';

type MaybePromise<T> = T | Promise<T>;
export type FieldValidator<Value, ValidationErrorCode: string> = (
  value: Value,
) => MaybePromise<FieldValidationResult<ValidationErrorCode>>;

export type FieldDescriptor<Value, ValidationErrorCode: string> = $ReadOnly<{
  initialValue: Value,
  validator?: FieldValidator<Value, ValidationErrorCode>,
  // validators?: $ReadOnlyArray<FieldValidator<Value, ErrorCode>>,
}>;

export type FormDescriptor = $ReadOnly<{
  // $FlowFixMe
  fields: { [key: string]: FieldDescriptor<any, any> },
}>;

export type Field<Value, ValidationErrorCode: string> = $ReadOnly<{|
  getState: () => FieldState<Value, ValidationErrorCode>,
  setValue: (value: Value) => void,
|}>;

export type FormData<Descriptor: FormDescriptor> = $ObjMap<
  $PropertyType<Descriptor, 'fields'>,
  // $FlowFixMe
  <Value>(FieldDescriptor<Value, any>) => Value,
>;

type FieldDescriptorFromDescriptor<Descriptor: FormDescriptor, FieldKey: string> = $ElementType<
  $PropertyType<Descriptor, 'fields'>,
  FieldKey,
>;

export type Form<Descriptor: FormDescriptor> = $ReadOnly<{|
  descriptor: Descriptor,
  getState: $PropertyType<FormStore<Descriptor>, 'getState'>,
  subscribe: $PropertyType<FormStore<Descriptor>, 'addConsumer'>,
  getField: <FieldKey: string>(
    fieldKey: FieldKey,
  ) => $Call<
    <Value, ValidationErrorCode: string>(
      FieldDescriptor<Value, ValidationErrorCode>,
    ) => Field<Value, ValidationErrorCode>,
    FieldDescriptorFromDescriptor<Descriptor, FieldKey>,
  >,
  getData: () => FormData<Descriptor>,
  validate: () => void,
|}>;

export default function createForm<Descriptor: FormDescriptor>(
  descriptor: Descriptor,
): Form<Descriptor> {
  const {
    getState,
    addConsumer: subscribe,
    updateState,
    getFieldState,
    updateFieldState,
  } = createFormStore(descriptor);

  const setFieldValidationResult = (fieldKey, validationResult) => {
    // $FlowFixMe
    updateFieldState(fieldKey, {
      isValid: typeof validationResult === 'string' ? false : validationResult,
      validationResult,
    });

    // Update form validity
    if (validationResult === null) {
      // $FlowFixMe
      updateState({ isValid: null });
    } else if (typeof validationResult === 'string') {
      // $FlowFixMe
      updateState({ isValid: false });
    } else {
      const isValid = Object.keys(descriptor.fields).reduce((prevIsValid, aFieldKey) => {
        if (prevIsValid !== true) return prevIsValid;

        return getFieldState(aFieldKey).isValid;
      }, true);

      // $FlowFixMe
      updateState({ isValid });
    }
  };

  const validateField = async (fieldKey) => {
    const { validator } = descriptor.fields[fieldKey];
    const fieldState = getFieldState(fieldKey);

    let validationResult = true;

    if (validator) {
      setFieldValidationResult(fieldKey, null);

      validationResult = await validator(fieldState.value);
    }

    // TODO: Do proper cancelation
    if (fieldState.value !== getFieldState(fieldKey).value) return;

    setFieldValidationResult(fieldKey, validationResult);
  };

  // $FlowFixMe
  const setFieldValue = (fieldKey: any, value: any) => {
    // $FlowFixMe
    updateFieldState(fieldKey, { value });
    validateField(fieldKey);
  };

  // $FlowFixMe
  const getField = (fieldKey): Field<any, any> => {
    const field = {
      getState: () => getFieldState(fieldKey),
      setValue: (value) => setFieldValue(fieldKey, value),
    };

    Object.freeze(field);

    return field;
  };

  const getData = () =>
    Object.keys(descriptor.fields).reduce((fv, fieldKey) => {
      // eslint-disable-next-line no-param-reassign
      fv[fieldKey] = getFieldState(fieldKey).value;

      return fv;
    }, {});

  const validate = () => {
    Object.keys(descriptor.fields).forEach(validateField);
  };

  const form: Form<Descriptor> = {
    descriptor,
    getState,
    getData,
    subscribe,
    getField,
    validate,
  };

  Object.freeze(form);

  return form;
}
