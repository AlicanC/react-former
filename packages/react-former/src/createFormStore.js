// @flow

import { createStore, type Store } from 'reductress-core';

import { type FormDescriptor, type FieldDescriptor } from './createForm';

export type Validity =
  | true // Valid
  | false // Invalid
  | null; // Pending

export type FieldValidationResult<ErrorCode: string> =
  | true // Valid
  | ErrorCode // Invalid
  | null; // Pending

export type FieldState<Value, ValidationErrorCode: string> = $ReadOnly<{
  value: Value,
  isValid: Validity,
  validationResult: FieldValidationResult<ValidationErrorCode>,
}>;

function createFieldState<Value, ValidationErrorCode: string>(
  fieldDescriptor: FieldDescriptor<Value, ValidationErrorCode>,
): FieldState<Value, ValidationErrorCode> {
  return {
    value: fieldDescriptor.initialValue,
    isValid: null,
    validationResult: null,
  };
}

type FormStateFields<Descriptor: FormDescriptor> = $ObjMap<
  $PropertyType<Descriptor, 'fields'>,
  typeof createFieldState,
>;

export type FormState<Descriptor: FormDescriptor> = $ReadOnly<{
  fields: FormStateFields<Descriptor>,
  isValid: Validity,
}>;

export type FormStore<Descriptor: FormDescriptor> = Store<FormState<Descriptor>> &
  $ReadOnly<{
    updateState: (formState: $Shape<FormState<Descriptor>>) => void,
    // $FlowFixMe
    getFieldState: <FieldKey: string>(fieldKey: FieldKey) => any,
    updateFieldState: <FieldKey: string>(
      fieldKey: FieldKey,
      // $FlowFixMe
      fieldState: $Shape<FieldState<any, any>>,
    ) => void,
  }>;

export default function createFormStore<Descriptor: FormDescriptor>(
  formDescriptor: Descriptor,
): FormStore<Descriptor> {
  const initialFieldStates = Object.keys(formDescriptor.fields).reduce((fv, fieldKey) => {
    const fieldDescription = formDescriptor.fields[fieldKey];

    // eslint-disable-next-line no-param-reassign
    fv[fieldKey] = createFieldState(fieldDescription);

    return fv;
  }, {});

  const initialState: FormState<Descriptor> = {
    fields: initialFieldStates,
    isValid: null,
  };

  const reductressStore = createStore(initialState);

  const updateState = (formState) => {
    const state = reductressStore.getState();

    // $FlowFixMe
    const nextState = {
      ...state,
      ...formState,
    };

    reductressStore.setState(nextState);
  };

  const getFieldState = (fieldKey) => reductressStore.getState().fields[fieldKey];
  const updateFieldState = (fieldKey, fieldState) => {
    const state = reductressStore.getState();

    // $FlowFixMe
    const nextState = {
      ...state,
      fields: {
        ...state.fields,
        [fieldKey]: {
          ...state.fields[fieldKey],
          // $FlowFixMe
          ...fieldState,
        },
      },
    };

    reductressStore.setState(nextState);
  };

  const formStore: FormStore<Descriptor> = {
    ...reductressStore,
    updateState,
    getFieldState,
    updateFieldState,
  };

  Object.freeze(formStore);

  return formStore;
}
