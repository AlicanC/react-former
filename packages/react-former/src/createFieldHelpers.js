// @flow

import { type Validity, type FieldValidationResult } from './createFormStore';
import { type Form, type FormDescriptor, type Field } from './createForm';

export type FieldHelper<Value, ValidationErrorCode: string> = $ReadOnly<{|
  value: Value,
  isValid: Validity,
  validationResult: FieldValidationResult<ValidationErrorCode>,
  setValue: (value: Value) => void,
  onChangeValue: (value: Value) => mixed,
  onChangeText: (value: Value) => mixed,
  // $FlowFixMe
  onChange: (event: any) => mixed,
|}>;

function createFieldHelper<Value, ValidationErrorCode: string>(
  field: Field<Value, ValidationErrorCode>,
): FieldHelper<Value, ValidationErrorCode> {
  const state = field.getState();

  const onChangeValue = (value) => field.setValue(value);

  return {
    value: state.value,
    isValid: state.isValid,
    validationResult: state.validationResult,
    setValue: field.setValue,
    onChangeValue: field.setValue,
    onChangeText: field.setValue,
    onChange: (event) => onChangeValue(event.target.value),
  };
}

type Fields<Descriptor: FormDescriptor> = $ObjMap<
  $PropertyType<Descriptor, 'fields'>,
  $PropertyType<Form<Descriptor>, 'getField'>,
>;
export type FieldHelpers<Descriptor: FormDescriptor> = $ObjMap<
  Fields<Descriptor>,
  typeof createFieldHelper,
>;

export default function createFieldHelpers<Descriptor: FormDescriptor>(
  form: Form<Descriptor>,
): FieldHelpers<Descriptor> {
  const fieldHelpers = Object.keys(form.descriptor.fields).reduce((helpers, fieldKey) => {
    const field = form.getField(fieldKey);

    // eslint-disable-next-line no-param-reassign
    helpers[fieldKey] = createFieldHelper(field);

    return helpers;
  }, {});

  return fieldHelpers;
}
