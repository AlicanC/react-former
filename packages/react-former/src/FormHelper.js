/* @flow */

import * as React from 'react';

import { type Form, type FormDescriptor } from './createForm';
import { type FormState } from './createFormStore';
import createFieldHelpers, { type FieldHelpers } from './createFieldHelpers';

export type Props<Descriptor: FormDescriptor> = {
  form: Form<Descriptor>,
  children: (fieldHelpers: FieldHelpers<Descriptor>, form: Form<Descriptor>) => React.Node,
  onFormStateChange?: (formState: FormState<Descriptor>) => void,
};

type State = {
  // $FlowFixMe
  fieldHelpers: FieldHelpers<any>,
};

export default class FormHelper<Descriptor: FormDescriptor> extends React.PureComponent<
  Props<Descriptor>,
  State,
> {
  state = {
    fieldHelpers: createFieldHelpers(
      // $FlowFixMe
      this.props.form,
    ),
  };
  // $FlowFixMe
  subscription: any;

  componentDidMount() {
    const { form } = this.props;

    this.subscription = form.subscribe((formState: FormState<Descriptor>) => {
      const { onFormStateChange } = this.props;

      if (onFormStateChange) {
        onFormStateChange(formState);
      }

      this.setState({
        fieldHelpers: createFieldHelpers(form),
      });
    });
  }

  componentWillUnmount() {
    if (this.subscription) {
      this.subscription.unsubscribe();
    }
  }

  render() {
    const { form, children: render } = this.props;
    const { fieldHelpers } = this.state;

    const children = render(
      // $FlowFixMe
      fieldHelpers,
      form,
    );

    return children;
  }
}
