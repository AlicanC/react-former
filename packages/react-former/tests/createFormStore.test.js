// @flow

import createFormStore from '../src/createFormStore';
import { type FormDescriptor } from '../src/createForm';

test('this is not enough', () => {
  const formDescriptor: FormDescriptor = {
    fields: {
      username: {
        initialValue: 'AlicanC',
      },
      password: {
        initialValue: 'pwisstronk',
      },
    },
  };

  const formStore = createFormStore(formDescriptor);

  expect(formStore.getState()).toMatchSnapshot();
});
