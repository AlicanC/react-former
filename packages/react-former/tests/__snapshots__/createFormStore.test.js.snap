// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`this is not enough 1`] = `
Object {
  "fields": Object {
    "password": Object {
      "isValid": null,
      "validationResult": null,
      "value": "pwisstronk",
    },
    "username": Object {
      "isValid": null,
      "validationResult": null,
      "value": "AlicanC",
    },
  },
  "isValid": null,
}
`;
